lib/cc/analyzer/path_filter.rb

```diff
@@ -0,0 +1,65 @@
+module CC
+  module Analyzer
+    class PathFilter
+      attr_reader :paths
+
+      def initialize(paths)
+        @paths = paths
+      end
+
+      def raise_if_any_unreadable_files
+        paths.each do |path|
+          if !File.directory?(path) && !FileUtils.readable_by_all?(path)
+            raise CC::Analyzer::UnreadableFileError, "Can't read #{path}"
+          end
+        end
+
+        self
+      end
+
+      def reject_unreadable_paths
+        @paths = paths.reject { |path| unreadable_path_entries.include?(path) }
+        self
+      end
+
+      def reject_paths(ignore_paths)
+        @paths = paths.reject { |path| ignore_paths.include?(path) }
+        self
+      end
+
+      def select_readable_files
+        @paths = paths.select { |path| FileUtils.readable_by_all?(path) }
+        self
+      end
+
+      def reject_symlinks
+        @paths = paths.reject { |path| File.symlink?(path) }
+        self
+      end
+
+      def reject_globs(globs)
+        @paths = paths.reject { |path| matches_globs?(path, globs) }
+        self
+      end
+
+      private
+
+      def unreadable_path_entries
+        @_unreadable_path_entries ||=
+          unreadable_paths.map { |path| PathEntries.new(path).entries }.flatten
+      end
+
+      def unreadable_paths
+        paths.select do |path|
+          File.directory?(path) && !FileUtils.readable_by_all?(path)
+        end
+      end
+
+      def matches_globs?(path, globs)
+        globs.any? do |exclude_path|
+          File.fnmatch(exclude_path, path)
```

Comment:
I just want to highlight that we've had a number of subtle bugs around some confusing behavior in Ruby.

Imagine the following structure:

```
foo.rb
foo/
  bar.rb
```

`Dir.glob("**/*.rb")` will find `foo.rb` (as expected). However, `File.fnmatch?("**/*.rb", "foo.rb")` is `false`.

This is why we (for a time) had to tell users they needed `**/*.rb` in excludes to "exclude all ruby files", but `**.rb` in ratings-paths to "rate all ruby file" -- confusing, and dumb.

So, BE CAREFUL. Where do `globs` come from? User input? Do you expect to see `**.rb` or `**/*.rb`? Is there a way to use the `PathPatterns` class, which we created in an effort to standardize on one kind of behavior in all our glob handling?